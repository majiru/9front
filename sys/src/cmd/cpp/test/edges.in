#define NOP(x) x
#define CAT(a, b) a ## b
#define EOF	(-1)

-- test nop --
x NOP(CAT(foo, EOF)) y
x NOP(CAT(EOF, foo)) y
x CAT(, EOF) y
y CAT(foo,) x
x CAT(,foo) y
X NOP(CAT(,)) y

-- test ncat --
#define NCAT(a)	foo ## a
NCAT(bar)

-- test xcat (no left arg) --
#define XCAT(a)	## a
foo XCAT(bar)

-- test cat3 --
#define CAT3(foo)	a##foo##b
CAT3(blah)

-- test expand and cat --
#define BAR	3
#define FOO	CAT(BAR, 3)
FOO

-- test expand and cat 2 --
/* Expected: a bc d */
CAT(a b, c d)
WUT

-- test varargs --
#define xprint(a, ...)	print(a, __VA_ARGS__)
xprint("hi", "there")
xprint("hi")

-- test expanding commas --
#define C	a,b
#define X(a)	a
#define Y	X(C)
Y

#define    x          3
#define    f(a)       f(x * (a))
#undef     x
#define    x          2
#define    g          f
#define    z          z[0]
#define    h          g(~
#define    m(a)       a(w)
#define    w          0,1
#define    t(a)       a
#define    p()        int
#define    q(x)       x
#define    r(x,y)     x ## y
#define    str(x)     # x
-- test complex expressions --
f(y+1) + f(f(z)) % t(t(g)(0) + t)(1);
g(x+(3,4)-w) | h 5) & m
(f)^m(m);

-- test empty args --
#define ZARGS(a) a b a
ZARGS()

-- test complex expressions with empty args --
p() i[q()] = { q(1), r(2,3), r(4,), r(,5), r(,) };
char c[2][6] = { str(hello), str() };
