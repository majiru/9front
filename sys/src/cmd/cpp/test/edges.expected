#line 1 "/sys/src/cmd/cpp/test/edges.out"
#line 1 "/sys/src/cmd/cpp/test/edges.in"




-- test nop --
x fooEOF y
x EOFfoo y
x(-1) y
y foo x
x foo y
X y

-- test ncat --

 foobar

-- test xcat (no left arg) --

foo ## bar

-- test cat3 --

 ablahb

-- test expand and cat --


 33

-- test expand and cat 2 --

 a bc d
WUT

-- test varargs --

 print("hi","there")
 print("hi",)

-- test expanding commas --



 a,b















-- test complex expressions --
 f(2 * (y+1)) + f(2 * (f(2 * (z[0])))) % f(2 * (0)) + t(1);
 f(2 * (2 +(3,4)- 0,1)) | f(2 * (~ 5)) & f(2 * (0,1))^ m(0,1);
#line 64 "/sys/src/cmd/cpp/test/edges.in"

-- test empty args --

 b

-- test complex expressions with empty args --
 int i[] = {1,23,4,5, };
char c[2][6] = {"hello","" };
